pyquil/numpy_simulator.py

```diff
@@ -0,0 +1,301 @@
+        """
+        if self.rs is None:
+            raise ValueError("You have tried to perform a stochastic operation without setting the "
+                             "random state of the simulator. Might I suggest using a PyQVM object?")
+
+        # note on reshape: it puts bitstrings in lexicographical order.
+        # would you look at that .. _all_bitstrings returns things in lexicographical order!
+        # reminder: qubit 0 is on the left in einsum simulator.
+        probabilities = np.abs(self.wf.reshape(-1)) ** 2
+        possible_bitstrings = all_bitstrings(self.n_qubits)
+        inds = self.rs.choice(2 ** self.n_qubits, n_samples, p=probabilities)
+        return possible_bitstrings[inds, :]
+
+    def do_measurement(self, qubit: int) -> int:
+        """
+        Measure a qubit, collapse the wavefunction, and return the measurement result.
+
+        :param qubit: Index of the qubit to measure.
+        :return: measured bit
+        """
+        if self.rs is None:
+            raise ValueError("You have tried to perform a stochastic operation without setting the "
+                             "random state of the simulator. Might I suggest using a PyQVM object?")
+
+        # Get probabilities
+        measurement_probs = get_measure_probabilities(self.wf, qubit)
+
+        # Flip a coin and record the result
+        measured_bit = int(np.argmax(self.rs.uniform() < np.cumsum(measurement_probs)))
+
+        # Zero out amplitudes corresponding to non-measured bistrings
+        other_bit = (measured_bit + 1) % 2
+        other_bit_indices = (slice(None),) * qubit + \
+                            (other_bit,) + \
+                            (slice(None),) * (self.n_qubits - qubit - 1)
+        self.wf[other_bit_indices] = 0
+
+        # Re-normalize amplitudes corresponding to measured bistrings
+        meas_bit_indices = (slice(None),) * qubit + \
+                           (measured_bit,) + \
+                           (slice(None),) * (self.n_qubits - qubit - 1)
+        self.wf[meas_bit_indices] /= np.sqrt(measurement_probs[measured_bit])
+        return measured_bit
+
+    def do_gate(self, gate: Gate):
+        """
+        Perform a gate.
+
+        :return: ``self`` to support method chaining.
+        """
+        gate_matrix, qubit_inds = _get_gate_tensor_and_qubits(gate=gate)
+        # TODO: choose between einsum and tensordot
+        # self.wf = targeted_einsum(gate=gate_matrix, wf=self.wf, wf_target_inds=qubit_inds)
+        self.wf = targeted_tensordot(gate=gate_matrix, wf=self.wf, wf_target_inds=qubit_inds)
+        return self
+
+    def expectation(self, operator: Union[PauliTerm, PauliSum]):
+        """
+        Compute the expectation of an operator.
+
+        :param operator: The operator
+        :return: The operator's expectation value
+        """
+        if not isinstance(operator, PauliSum):
+            operator = PauliSum([operator])
+
+        sum = 0
```

Comment:
`return sum([_term_expectation(self.wf, term) for term in operator])`

Reply:
good call. Don't know why I didn't do this

Reply:
you don't even need to cons up the list, right?

```
sum(_term_exp(...) for term in operator)
```